libsolidity/analysis/TypeChecker.cpp

```diff
@@ -583,13 +583,18 @@ bool TypeChecker::visit(VariableDeclaration const& _variable)
 
 	if (auto referenceType = dynamic_cast<ReferenceType const*>(varType))
 	{
-		auto result = referenceType->validForLocation(referenceType->location());
+		BoolResult result = referenceType->validForLocation(referenceType->location());
 		if (result)
 		{
 			bool isLibraryStorageParameter = (_variable.isLibraryFunctionParameter() && referenceType->location() == DataLocation::Storage);
 			bool callDataCheckRequired = ((_variable.isConstructorParameter() || _variable.isPublicCallableParameter()) && !isLibraryStorageParameter);
 			if (callDataCheckRequired)
-				result = referenceType->validForLocation(DataLocation::CallData);
+			{
+				if (!referenceType->interfaceType(false))
+					solAssert(m_errorReporter.hasErrors(), "");
+				else
+					result = referenceType->validForLocation(DataLocation::CallData);
+			}
```

Comment:
I don't know what the logic here is. My understanding is that we used to invoke `validForLocation(DataLocation::CallData` meaning that we always checked if it is okay to store the reference type in calldata, for constructor and public variables (I suppose function parameter counts as a public variable). If it couldn't be stored in calldata, we reported an error.

Now, we invoke `validForLocation` only if referenceType->interfaceType(false) is true which means only if the interface type of the variable (that is not in library) exists.

A few questions come up:

- Since we assume `interfaceType(false)`, is there a scenario where validForLocation(DataLocation::CallData) is called without a prior check for `interfaceType(true)`?
- When does a reference type not have an interface type? Presumably when the type is somehow not public exposed? What happens with
- 
```
pragma abicoder v2;

contract C {
    struct S { function() internal a; function() external b; }
    function f(S[2] memory) public {}
}
```

Reply:
The struct does not have an interface type because one of its members does not have an interface type (the internal function). The current logic works because there is another earlier check that looks for a non-interface type. That's why the assertion that checks that there has been a previous error is not failing.